WallsAndHoles/tilemap.cpp

```diff
@@ -97,21 +102,67 @@ void TileMap::resizeMap(QSize newSize)
     emit resized();
 }
 
-void TileMap::updateDepend(){
-    QVector<bool> valid(mDependencies.size());
-    for (int x = 0; x < mMap.size().width(); ++x) {
-        for (int y = 0; y < mMap.size().height(); ++y) {
-            for(int i=0; i<mDependencies.size(); i++){
-                if(mDependencies[i]->cTileTemplates().indexOf(mMap(x, y)->tileTemplate())>=0){
-                    valid[i]=true;
-                }
-            }
-        }
+bool TileMap::tileTemplateUsed(TileTemplate *tileTemplate)
+{
+    if (!tileTemplate) return false;
+
+    mTilePingReceiveMode = SetCheck;
+    mTilePinged = false;
+
+    tileTemplate->emitTilePing();
+
+    bool result = mTilePinged;
+    mTilePinged = false;
+    mTilePingReceiveMode = None;
+
+    return result;
+}
+
+bool TileMap::tileTemplateSetUsed(TileTemplateSet *tileTemplateSet)
+{
+    mTilePingReceiveMode = SetCheck;
+    mTilePinged = false;
+
+    for (TileTemplate *t : tileTemplateSet->cTileTemplates()) {
+        if (t)
+            t->emitTilePing();
+
+        if (mTilePinged)
+            break;
     }
-    QVector<SharedTileTemplateSet> newDependencies;
-    for(int i=0; i<valid.size(); i++){
-        if(valid[i])
-            newDependencies.push_back(mDependencies[i]);
+
+    bool result = mTilePinged;
+    mTilePinged = false;
+    mTilePingReceiveMode = None;
+
+    return result;
+}
+
+void TileMap::removingTileTemplateSet(TileTemplateSet *tileTemplateSet)
+{
+    mTilePingReceiveMode = Collect;
+    mPingedTiles.clear();
+
+    for (TileTemplate *t : tileTemplateSet->cTileTemplates())
+        if (t)
+            t->emitTilePing();
```

Comment:
Cool system! I recommend using a QMutex whenever you do a ping --- Qt makes a lot of threads, and it's totally possible that if you have two methods that use the ping system, they may be called simultaneously and corrupt each other. Usage is simple:

In tilemap.h:
```
QMutex mPingingMutex;
```

in tilemap.cpp, wherever needed:
```
mPingingMutex.lock(); // If the mutex is already locked, this method will block until it is unlocked.
// Pinging code here
mPingingMutex.unlock();
```

Reply:
Also, beware that improper use can cause a deadlock. If the mutex is locked, do not call any methods that may try to lock it again (at least not on the same thread)!